patches/tock/01-persistent-storage.patch

```diff
@@ -285,39 +313,41 @@ index 5abd2d84..5a726fdb 100644
 +        }
 +    }
 +}
-diff --git a/kernel/src/callback.rs b/kernel/src/callback.rs
-index c812e0bf..bd1613b3 100644
---- a/kernel/src/callback.rs
-+++ b/kernel/src/callback.rs
-@@ -130,6 +130,31 @@ impl AppId {
-             (start, end)
-         })
-     }
-+
-+    pub fn in_writeable_flash_region(&self, ptr: usize, len: usize) -> bool {
-+        self.kernel.process_map_or(false, *self, |process| {
-+            let ptr = match ptr.checked_sub(process.flash_start() as usize) {
-+                None => return false,
-+                Some(ptr) => ptr,
-+            };
-+            (0..process.number_writeable_flash_regions()).any(|i| {
-+                let (region_ptr, region_len) = process.get_writeable_flash_region(i);
-+                let region_ptr = region_ptr as usize;
-+                let region_len = region_len as usize;
-+                // We want to check the 2 following inequalities:
-+                // (1) `region_ptr <= ptr`
-+                // (2) `ptr + len <= region_ptr + region_len`
-+                // However, the second one may overflow written as is. We introduce a third
-+                // inequality to solve this issue:
-+                // (3) `len <= region_len`
-+                // Using this third inequality, we can rewrite the second one as:
-+                // (4) `ptr - region_ptr <= region_len - len`
-+                // This fourth inequality is equivalent to the second one but doesn't overflow when
-+                // the first and third inequalities hold.
-+                region_ptr <= ptr && len <= region_len && ptr - region_ptr <= region_len - len
-+            })
-+        })
-+    }
- }
+diff --git a/kernel/src/process.rs b/kernel/src/process.rs
+index eb00f274..663c2422 100644
+--- a/kernel/src/process.rs
++++ b/kernel/src/process.rs
+@@ -1604,6 +1604,33 @@ impl<C: 'static + Chip> Process<'a, C> {
+             return Ok((None, 0));
+         }
  
- /// Type to uniquely identify a callback subscription across all drivers.
++        // Allocate MPU region for storage. The storage is currently static and readable by all
++        // processes. This will be fixed once Tock supports persistent storage outside the
++        // application flash (i.e. not the current writeable flash regions support).
++        const STORAGE_PTR: usize = 0xc0000;
```

Comment:
The change is much better with the address and size only being put in the kernel.
The main problem now is that we can't even compile/run a Tock kernel that is not for the nRF52840 chip because of this address. I still think this would be better to have it in the `nrf52840_layout.ld`. Just like there is a `FLASH` and a `SRAM` portion, we should be able to define a `STORAGE` area and use the symbols in the kernel directly. The tricky part (I'm not familiar enough with ldscript files) is to handle the cases where the area doesn't exist.
Another approach that is being used in `libtock-rs` is to use environment variables at compile time. This might be less complicated and would allow to handle the case where the kernel doesn't need a storage area. WDYT?

Reply:
Oh I see the problem. I didn't anticipate that the MPU part would be an issue for other boards. The part in NVMC is fine because it's only for nrf52.

Indeed, trying to use the linker script through the TBF could be a solution. I need to look into it. I might have seen something relevant in Tock on Titan about that. I need to check again.

I guess the environment variable should be a last resort solution right?

Reply:
As you said, it's for `nrf52` which is a big family and `nrf52840` is the biggest one in this chip family in terms of flash. So having a hard-coded address is likely to cause a kernel panic at runtime if the address doesn't exist because the flash is smaller.

Changing the linker script is one option. As I said, it might be tricky to handle a non-existing symbol. This may require us to define the symbol for all the boards. But on the other hand, it may allow us to do some sanity checks that the defined area for storage will work with the MPU (it has the [constraint](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dai0179b/CHDFDFIG.html) that the base address of the region must be aligned to a multiple of its size).

The env variable gives us more flexibility and we fully control the environment with the `deploy.py` script. That's probably a quickest option that will let us investigate the linker script road at a later stage.

Reply:
I don't see a way to use the linker script without having to change elf2tab. I see those 2 solutions (assuming we want to define the storage in the linker script):
- We change the semantics of writable flash regions to actually represent the out-of-app persistent storage. We probably don't need to change much on kernel side because writable flash regions don't have extensive support (but we would still need to review where it's used). We need to change elf2tab to read the storage pointer and size in the ELF and output them as a writable flash region.
- We add a new TbfHeaderType for persistent storage. We implement the parsing of this header in the Kernel. We implement the generation of this header in elf2tab and populate it from the ELF.

In both cases, we need a way to output the storage address and size in the ELF. I guess we could use a section for that. We would filter out this section in elf2tab and only use it to know how to write the TBF headers.

I think the env variable looks actually like a much simpler approach, unless we find a way that doesn't need to modify elf2tab.

Reply:
The last commit defines the storage locations in the board, which doesn't need to modify elf2tab as it would be the case with the linker script. So the storage locations are now defined in a single place which is board-specific.

I also added support for cases where the persistent storage spans multiple non-contiguous locations, like it could be the case if the flash had multiple banks and one would want to use the last pages of each bank for the persistent storage.

The diff between this state and possible future Tock states is:
- The persistent storage is readable by all apps. For that we would need stable app ids or define the persistent storage in the app (which needs to modify elf2tab).
- The persistent storage is statically defined by the board. It could instead be defined by metadata in the list of apps and thus dynamically edited by tockloader. It could also be defined inside each app if support for multiple apps accessing the same persistent storage is not needed.

Remaining to do: Use the syscall filter to prevent the application from mutating the flash outside the persistent storage. This test is currently done by the driver.

Reply:
It's actually not possible to use the syscall filter because we can't access the driver state from the syscall filter. We need to know the length of the allowed slice. I'll try to see if there is some issue regarding syscall filter to mention this problem.

Reply:
I think this hook is a first step at filtering syscall but there are more PRs planned around that. We can revisit the syscall filter at a later stage

Reply:
By the way, I just asked [here](https://github.com/tock/tock/issues/1350#issuecomment-628023606) whether this would be supported in the future.

Reply:
To move the storage test from the driver to the syscall driver, I had to change the driver API in the following ways:
- Move the storage locations accessors from the driver to `memop` syscalls.
- Add a length argument to the write slice and erase page commands. This is redundant but works around the current syscall filter limitation.

Now that the NVMC driver only uses the app to access the grants and not to access the writable flash regions, I'll try to resuscitate https://github.com/tock/tock/pull/1467. If accepted it would simplify a bit the persistent storage patch we have to only address the future persistent storage support of Tock.